File: solutions/problem_2171_removing_minimum_number_of_magic_beans/solution.py
"""
Problem 2171: Removing Minimum Number of Magic Beans
==================================================
Difficulty: Medium
Tags: Array, Greedy, Sorting, Enumeration, Prefix Sum
Pattern: Greedy with Sorting

Time Complexity:  O(n log n)
Space Complexity: O(n)
"""

class Solution:
    def minimumRemoval(self, beans: List[int]) -> int:
        # Sort + greedy - O(n log n) time
        beans.sort()
        result = 0
        curr_end = 0
        for item in beans:
            if isinstance(item, (list, tuple)):
                if item[0] >= curr_end:
                    result += 1
                    curr_end = item[1]
            else:
                result += 1
        return result

File: solutions/problem_2171_removing_minimum_number_of_magic_beans/solution.cpp
/*
 * Problem 2171: Removing Minimum Number of Magic Beans
 * ==================================================
 * Difficulty: Medium
 * Tags: Array, Greedy, Sorting, Enumeration, Prefix Sum
 * Pattern: Greedy with Sorting
 *
 * Time Complexity:  O(n log n)
 * Space Complexity: O(n)
 */

#include <algorithm>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
    int minimumRemoval(vector<int>& beans) {
        // Sort + greedy - O(n log n) time
        sort(beans.begin(), beans.end());
        int result = 0, curr_end = 0;
        for (auto& item : beans) {
            result++;
        }
        return result;
    }
};
